File: server/app/schemas/parse.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.models.enums import ScheduleSource


ScheduleStorageStrategy = Literal[
    "local_only",
    "sync_to_cloud",
    "sync_to_cloud_and_knowledge",
]
ParseAgentAction = Literal["ask_follow_up", "finalize_draft"]
ParseAgentToolName = Literal["update_draft", "ask_follow_up", "finalize_draft", "save_schedule_to_local"]
ParseAgentMessageRole = Literal["user", "assistant"]


class ParseDraftRequest(BaseModel):
    text: str = Field(min_length=1, max_length=4000)
    reference_time: datetime | None = None


class ScheduleDraft(BaseModel):
    title: str | None = None
    start_time: datetime | None = None
    end_time: datetime | None = None
    location: str | None = None
    remark: str | None = None
    source: ScheduleSource = ScheduleSource.AI_PARSED
    storage_strategy: ScheduleStorageStrategy | None = None


class ScheduleDraftPatch(BaseModel):
    title: str | None = Field(default=None, max_length=200)
    start_time: datetime | None = None
    end_time: datetime | None = None
    location: str | None = Field(default=None, max_length=255)
    remark: str | None = None
    storage_strategy: ScheduleStorageStrategy | None = None


class ParseFollowUpQuestion(BaseModel):
    field: str
    question: str


class ParseDraftResponse(BaseModel):
    draft: ScheduleDraft
    missing_fields: list[str]
    follow_up_questions: list[ParseFollowUpQuestion] = Field(default_factory=list)
    requires_human_review: bool = True
    can_persist_directly: bool = False


class ParseSessionCreateRequest(BaseModel):
    message: str = Field(min_length=1, max_length=4000)
    reference_time: datetime | None = None


class ParseSessionMessageRequest(BaseModel):
    message: str = Field(min_length=1, max_length=4000)
    reference_time: datetime | None = None


class ParseSessionDraftPatchRequest(BaseModel):
    draft: ScheduleDraftPatch


class ParseAgentMessage(BaseModel):
    id: str
    role: ParseAgentMessageRole
    content: str


class ParseAgentToolCall(BaseModel):
    name: ParseAgentToolName
    summary: str


class ParseSessionResponse(BaseModel):
    parse_session_id: str
    messages: list[ParseAgentMessage]
    draft: ScheduleDraft
    missing_fields: list[str]
    follow_up_questions: list[ParseFollowUpQuestion] = Field(default_factory=list)
    ready_for_confirm: bool
    next_action: ParseAgentAction
    tool_calls: list[ParseAgentToolCall] = Field(default_factory=list)
    latest_assistant_message: str | None = None
    draft_visible: bool = False

File: frontend/src/api/parse.ts
import { api } from "@/api/client";
import type { ScheduleStorageStrategy } from "@/repositories/local-schedules";

const PARSE_REQUEST_TIMEOUT_MS = 45_000;

export type ParseScheduleDraft = {
  title: string | null;
  start_time: string | null;
  end_time: string | null;
  location: string | null;
  remark: string | null;
  source: "ai_parsed";
  storage_strategy: ScheduleStorageStrategy | null;
};

export type ParseAgentMessage = {
  id: string;
  role: "user" | "assistant";
  content: string;
};

export type ParseAgentTraceEntry = {
  action:
    | "build_context"
    | "plan_update"
    | "apply_draft_update"
    | "request_clarification"
    | "prepare_confirmation";
  summary: string;
  source?: "runtime" | "heuristic" | "manual_patch" | null;
};

export type ParseAgentToolCall = {
  name: "update_draft" | "ask_follow_up" | "finalize_draft" | "save_schedule_to_local";
  summary: string;
};

export type ParseDraftResponse = {
  draft: ParseScheduleDraft;
  missing_fields: string[];
  follow_up_questions: Array<{ field: string; question: string }>;
  state: "clarifying" | "ready_for_confirm";
  trace: ParseAgentTraceEntry[];
  requires_human_review: boolean;
  can_persist_directly: boolean;
};

export type ParseSessionResponse = {
  parse_session_id: string;
  messages: ParseAgentMessage[];
  draft: ParseScheduleDraft;
  missing_fields: string[];
  follow_up_questions: Array<{ field: string; question: string }>;
  state: "clarifying" | "ready_for_confirm";
  ready_for_confirm: boolean;
  next_action: "ask_follow_up" | "finalize_draft";
  tool_calls: ParseAgentToolCall[];
  trace: ParseAgentTraceEntry[];
  latest_assistant_message: string | null;
  draft_visible: boolean;
};

export type ParseSessionCreateRequest = {
  message: string;
  reference_time: string;
};

export type ParseSessionMessageRequest = {
  message: string;
  reference_time: string;
};

export type ParseSessionDraftPatchRequest = {
  draft: {
    title?: string | null;
    start_time?: string | null;
    end_time?: string | null;
    location?: string | null;
    remark?: string | null;
    storage_strategy?: ScheduleStorageStrategy | null;
  };
};

export async function createParseSession(payload: ParseSessionCreateRequest): Promise<ParseSessionResponse> {
  const response = await api.post<ParseSessionResponse>("/parse/sessions", payload, {
    timeout: PARSE_REQUEST_TIMEOUT_MS
  });
  return response.data;
}

export async function continueParseSession(
  sessionId: string,
  payload: ParseSessionMessageRequest
): Promise<ParseSessionResponse> {
  const response = await api.post<ParseSessionResponse>(`/parse/sessions/${sessionId}/messages`, payload, {
    timeout: PARSE_REQUEST_TIMEOUT_MS
  });
  return response.data;
}

export async function patchParseSessionDraft(
  sessionId: string,
  payload: ParseSessionDraftPatchRequest
): Promise<ParseSessionResponse> {
  const response = await api.patch<ParseSessionResponse>(`/parse/sessions/${sessionId}/draft`, payload, {
    timeout: PARSE_REQUEST_TIMEOUT_MS
  });
  return response.data;
}
